Rewrite the section about dynamic updates (#117)
AMWA-TV/nmos-sink-metadata-processing: docs/Behaviour - Client Side.md

Constraint Sets of Receiver Capabilities with `urn:x-nmos:cap:meta:enabled` set to false MUST be ignored completely while making the processing.

## Dynamic Update of Resources

> The Controller MUST be capable of using Stream Compatibility API in conjunction with IS-04 Query API (if available) or IS-04 Node API to discover and dynamically report the state of any relevant resources, including Inputs, Outputs, Senders and Receivers.
> 

Author: ```suggestion
The Controller MUST be capable of using Stream Compatibility API in conjunction with IS-04 APIs to discover and dynamically report the state of any relevant resources, including Inputs, Outputs, Senders and Receivers.

In order to avoid polling of the Stream Compatibility API, it is RECOMMENDED that Controllers use the WebSocket subscription mechanism of IS-04 Query API.
```

The Controller MUST be capable of using Stream Compatibility API in conjunction with IS-04 Query API (if available) or IS-04 Node API to discover and dynamically report the state of any relevant resources, including Inputs, Outputs, Senders and Receivers.

> NMOS Controller MUST track Sender's `subscription`'s `active` property during and after the IS-05 connection. If it became `false` unexpectedly, NMOS Controller SHOULD `GET /status` to figure out the reason of breaking the connection. If the State of the Sender is `active_constraints_violation`, then NMOS Controller SHOULD either disconnect all the Receivers or `PUT /constraints/active` once again and re-enable Sender via IS-05.

Reviewer: You removed the "tracking sender subscription active property" aspect ... We must favor monitoring through the registry query/subscription mechanism first ... I don't think it would be a good idea for Controller to poll the status/state. In all aspects of Controller monitoring we must make sure that it is being performed not by polling the device but the registry.

Author: If I correctly understood Gareth in #116, it's implied that Controller tracks resources' properties, we need to dictate what Controller should do with their values. Moreover, `active` property update is a consequence of changing the State of a Sender to `active_constraints_violation` so maybe this is what Controller should track instead.

[The current text](https://specs.amwa.tv/is-11/branches/v1.0-dev/docs/Interoperability.html#version-increments) says:
>When State of Sender or Receiver is changed, or when the set of Inputs/Outputs associated with the Sender/Receiver is changed, then the version attribute of the relevant IS-04 Sender or Receiver MUST be incremented.

Do I correctly understand, that it means that Controller doesn't have to periodically poll `/status` and can it do that only after gets `version` update from Query API?

>In all aspects of Controller monitoring we must make sure that it is being performed not by polling the device but the registry.

I tried to express it with `The Controller MUST be capable of using Stream Compatibility API in conjunction with IS-04 Query API (if available) or IS-04 Node API`. [IS-04 RECOMMENDs](https://specs.amwa.tv/is-04/releases/v1.3.2/docs/Controllers.html#websocket-subscriptions) to use WebSocket subscriptions. Inserted this paragraph as a suggestion.

Reviewer: "Do I correctly understand, that it means that Controller doesn't have to periodically poll /status and can it do that only after gets version update from Query API?"

What the text says is a requirement for the Node.

"In all aspects of Controller monitoring we must make sure that it is being performed not by polling the device but the registry."

Where is this sentence coming from?

At the IS-11 level we must make sure people will not mis-interpret our specification because they haven't read the full IS-04 and/or Controller specifications and emphasis when possible that because an IS-11 Node have a requirement to update the Sender's version is various cases, the Controller SHOULD monitor from the Registry when available.

> * The Controller MUST indicate statutes of Inputs, Outputs, Senders and Receivers.

Reviewer: "statutes" ???

There are a number of new MUST clauses. Are there any other API requiring the Controller to "indicate" the state of something?

Reviewer 2: Let's composer with the language in e.g. https://specs.amwa.tv/is-05/releases/v1.1.2/docs/Controllers.html#connection-management

("statuses" I guess!)

Author: ```suggestion
* The Controller MUST indicate statuses of Inputs, Outputs, Senders and Receivers.
```

Constraint Sets of Receiver Capabilities with `urn:x-nmos:cap:meta:enabled` set to false MUST be ignored completely while making the processing.

## Dynamic Update of Resources

The Controller MUST be capable of using Stream Compatibility API in conjunction with IS-04 APIs to discover and dynamically report the state of any relevant resources, including Inputs, Outputs, Senders and Receivers.

In order to avoid polling of the Stream Compatibility API, it is RECOMMENDED that Controllers use the WebSocket subscription mechanism of IS-04 Query API.

> * The Controller MUST indicate statuses of Inputs, Outputs, Senders and Receivers.

Reviewer: If there is no inputs, outputs the Controller does not have any requirement, not even to indicate that there is none. Same thing when no input are associated to a Sender and no output is associated to a Receiver ... For Nodes not supporting the concept of input/output we must make sure the Controller has no requirement about something not supported by the Node.

Author: Tried to cover it in the suggestion below.

* The Controller MUST report to the User when `status` attribute of the state of a Sender becomes `active_constraints_violation` and SHOULD report when it becomes `no_essence` or `awaiting_essence`.
* The Controller SHOULD report to the User when `status` attribute of the state of an active Receiver becomes `non_compliant_stream`.
* The Controller SHOULD `PATCH /staged` of all the connected Receivers with the new `transport_file` requesting immediate activation when the related active Sender updates the transport file.

> * The Controller MUST reflect changes in presence/absence of Inputs and Outputs to the User after a maximum of 30 seconds.

Reviewer: Are you referring to Resources appearing in the /inputs, /outputs, /senders/*inputs or /receivers/*/outputs (id come and go) or to the state (connected, no_signal, ...)?

Author: The last item here is about showing Inputs/Outputs presence (`/inputs/{inputId}/properties` and `/outputs/{outputId}/properties`). A Controller, that shows only IDs coming and going, satisfies this requirement. Showing states is covered by the item before the first item here (not included to your quote but it's the 1st in this list).

There is no requirement about showing relationship between Inputs/Outputs with Senders/Receivers though. Maybe we have to add it as well.

> * The Controller MUST indicate statuses of Inputs, Outputs, Senders and Receivers.
> * The Controller MUST report to the User when `status` attribute of the state of a Sender becomes `active_constraints_violation` and SHOULD report when it becomes `no_essence` or `awaiting_essence`.
> * The Controller SHOULD report to the User when `status` attribute of the state of an active Receiver becomes `non_compliant_stream`.
> * The Controller SHOULD `PATCH /staged` of all the connected Receivers with the new `transport_file` requesting immediate activation when the related active Sender updates the transport file.
> * The Controller MUST reflect changes in presence/absence of Inputs and Outputs to the User after a maximum of 30 seconds.

Author: ```suggestion
* The Controller MUST indicate statuses of Inputs, Outputs, Senders and Receivers and relationship (or its absence) between Inputs/Outputs and Senders/Receivers listed in Stream Compatibility APIs.
* The Controller MUST indicate whether each Sender/Receiver discovered via IS-04 APIs is present in Stream Compatibility API
* The Controller MUST report to the User when `status` attribute of the state of a Sender becomes `active_constraints_violation` and SHOULD report when it becomes `no_essence` or `awaiting_essence`.
* The Controller SHOULD report to the User when `status` attribute of the state of an active Receiver becomes `non_compliant_stream`.
* The Controller SHOULD `PATCH /staged` of all the connected Receivers with the new `transport_file` requesting immediate activation when the related active Sender updates the transport file.
* The Controller MUST reflect changes in presence/absence of Inputs and Outputs to the User after a maximum of 30 seconds.
```

Reviewer: awaiting_essence is a transient state and MUST not be reported to the user.

Reviewer: "when `status` attribute of the state of a Sender" => seems to be "when `state` attribute of the status of a Sender"

Author: ```suggestion
* The Controller MUST indicate statuses of Inputs, Outputs, Senders and Receivers and relationship (or its absence) between Inputs/Outputs and Senders/Receivers listed in Stream Compatibility APIs.
* The Controller MUST indicate whether each Sender/Receiver discovered via IS-04 APIs is present in Stream Compatibility API
* The Controller MUST report to the User when `state` attribute of the status of a Sender becomes `active_constraints_violation` and SHOULD report when it becomes `no_essence`.
* The Controller SHOULD report to the User when `state` attribute of the status of an active Receiver becomes `non_compliant_stream`.
* The Controller SHOULD `PATCH /staged` of all the connected Receivers with the new `transport_file` requesting immediate activation when the related active Sender updates the transport file.
* The Controller MUST reflect changes in presence/absence of Inputs and Outputs to the User after a maximum of 30 seconds.
```
Removed `awaiting_essence`. Should we dictate that Controller MUST not report about it?

Reviewer: "The Controller MUST indicate statuses of Inputs, Outputs, Senders and Receivers and relationship (or its absence) between Inputs/Outputs and Senders/Receivers listed in Stream Compatibility APIs."

It seems to me that "relationship" is not precise enough. I would phrase it as follow instead:

"The Controller MUST indicate statuses of Inputs, Outputs, Senders and Receivers and SHOULD indicate the Inputs associated to a given Sender and Outputs associated to a given Receiver for Senders/Receivers listed in Stream Compatibility APIs."

The word "relationship" in general could have meant to show the interconnection of inputs with senders in general and outputs with receivers in general and I think that this is out of scope for IS-11.

Reviewer: "* The Controller MUST report to the User when `state` attribute of the status of a Sender becomes `active_constraints_violation` and SHOULD report when it becomes `no_essence`."

I would be careful here and not use MUST here but a SHOULD because of the "volatility" of the `active_constraints_violation` state .. For example when changing the EDID preferred mode the source does not react instantaneously such that the transition may not be clean and the Sender may get to the `active_constraints_violation` state. The Controller in such cases SHOULD restart the Sender automatically without telling the User because it may be a temporary state and a subsequent restart would finally remain stable with the state `constrained`.

To me the following text would be more appropriate.

"* The Controller SHOULD report to the User when `state` attribute of the status of a Sender becomes `active_constraints_violation`  or `no_essence`."

Otherwise this would also be acceptable:

"* The Controller MUST report to the User when `state` attribute of the status of a Sender remains `active_constraints_violation` or  `no_essence` after a maximum of 30 seconds."

Author: ```suggestion
* The Controller MUST indicate whether each Sender/Receiver discovered via IS-04 APIs is present in Stream Compatibility API and indicate statuses of such Senders/Receivers
* The Controller MUST report to the User when `state` attribute of the status of a Sender remains `active_constraints_violation` or `no_essence` after a maximum of 30 seconds.
* The Controller SHOULD report to the User when `state` attribute of the status of an active Receiver becomes `non_compliant_stream`.
* The Controller SHOULD be capable of performing `PATCH /staged` of all the connected Receivers with the new `transport_file` requesting immediate activation when the related active Sender updates the transport file.
* The Controller SHOULD indicate statuses of Inputs and Outputs and indicate the Inputs associated to a given Sender and Outputs associated to a given Receiver for Senders/Receivers listed in Stream Compatibility APIs.
* The Controller SHOULD reflect changes in presence/absence of Inputs and Outputs to the User after a maximum of 30 seconds.
```
Added the 30 seconds timeout for reporting about `state` of a Sender.

Author: >The Controller MUST indicate statuses of Inputs, Outputs, Senders and Receivers and SHOULD indicate the Inputs associated to a given Sender and Outputs associated to a given Receiver for Senders/Receivers listed in Stream Compatibility APIs.

It may be read as if orphan Inputs/Outputs can be ignored at all. Does the following sound more precise?
>The Controller MUST indicate statuses of Inputs, Outputs, Senders and Receivers listed in Stream Compatibility APIs and SHOULD indicate Inputs associated to a given Sender and Outputs associated to a given Receiver for such Senders/Receivers.

Constraint Sets of Receiver Capabilities with `urn:x-nmos:cap:meta:enabled` set to false MUST be ignored completely while making the processing.

## Dynamic Update of Resources

> The Controller MUST be capable of using Stream Compatibility API in conjunction with IS-04 APIs to discover and dynamically report the state of any relevant resources, including Inputs, Outputs, Senders and Receivers.

Author: ```suggestion
The Controller uses Stream Compatibility API in conjunction with IS-04 APIs to discover and track the state of any relevant resources, including Inputs, Outputs, Senders and Receivers.
```

Reviewer: Should it be more specific "The Controller uses Stream Compatibility API in conjunction with IS-04 Query API to discover and track the state of any relevant resources, including Inputs, Outputs, Senders and Receivers." because IS-04 says "The Controller uses the Query API to discover and track the state of any registered resource, including Node, Device, Source, Flow, Sender, and Receiver."

Author: There is no Query API in the Peer to Peer mode and I tried to include this case. I will raise it in the IS-04 PR.

> In order to avoid polling of the Stream Compatibility API, it is RECOMMENDED that Controllers use the WebSocket subscription mechanism of IS-04 Query API.

Author: Should we align this text with https://specs.amwa.tv/is-05/releases/v1.1.2/docs/Controllers.html#version-timestamp?

> In order to avoid polling of the Stream Compatibility API, it is RECOMMENDED that Controllers use the WebSocket subscription mechanism of IS-04 Query API.

Author: ```suggestion
```

* The Controller MUST indicate whether each Sender/Receiver discovered via IS-04 APIs is present in Stream Compatibility API and indicate statuses of such Senders/Receivers
* The Controller MUST report to the User when `state` attribute of the status of a Sender remains `active_constraints_violation` or `no_essence` after a maximum of 30 seconds.
* The Controller SHOULD report to the User when `state` attribute of the status of an active Receiver becomes `non_compliant_stream`.

> * The Controller SHOULD be capable of performing `PATCH /staged` of all the connected Receivers with the new `transport_file` requesting immediate activation when the related active Sender updates the transport file.

Author: This requires the Controller to act on events other than user input. `SHOULD` may be too strong for it.

Reviewer: I'm not sure I understand which SHOULD is too strong ... could you be more precise?

Author: The last one.
